File: aoc-2024-cpp/day10/day10.cpp
#include <algorithm>
#include <array>
#include <chrono>
#include <concepts>
#include <cstddef>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <iterator>
#include <map>
#include <optional>
#include <ranges>
#include <regex>
#include <set>
#include <string>
#include <string_view>
#include <vector>

#include <fmt/base.h>
#include <fmt/ranges.h>
#include <flux.hpp>

namespace sr = std::ranges;
namespace sv = std::views;
namespace fs = std::filesystem;

using fmt::print;
using fmt::println;

#define MEASURE(f)                                                             \
  {                                                                            \
    auto start = std::chrono::high_resolution_clock::now();                    \
    f;                                                                         \
    auto end = std::chrono::high_resolution_clock::now();                      \
    println("{} took {} µs", #f,                                               \
            std::chrono::duration_cast<std::chrono::microseconds>(end - start) \
                .count());                                                     \
  }

auto read_input(fs::path const& file_name) -> std::vector<std::string> {
  std::fstream infile(fs::path{BASE_PATH} / file_name);
  std::vector<std::string> lines;

  std::string line;
  while (std::getline(infile, line)) {
    lines.push_back(line);
  }
  infile.close();
  return lines;
}

constexpr static std::array<std::pair<int, int>, 4> directions = {
    {{-1, 0}, {0, 1}, {1, 0}, {0, -1}}};

void part1(std::vector<std::vector<int>> const& topo) {
  const auto rows = topo.size();
  const auto cols = topo[0].size();
  std::size_t score = 0;

  auto trace_path = [&](std::size_t r, std::size_t c) {
    std::vector<std::pair<int, std::pair<int, int>>> trails;
    std::set<std::pair<int, int>> destinations;

    trails.push_back({0, {r, c}});

    while (!trails.empty()) {
      auto [trail_val, trail_pos] = trails.back();
      trails.pop_back();
      auto [tr, tc] = trail_pos;
      for (auto [dr, dc] : directions) {
        auto nr = tr + dr;
        auto nc = tc + dc;
        if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) {
          continue;
        }
        if (trail_val == 8 && topo[nr][nc] == 9) {
          destinations.insert({nr, nc});
          continue;
        }
        if (topo[nr][nc] == trail_val + 1) {
          trails.push_back({trail_val + 1, {nr, nc}});
        }
      }
    }
    auto trail_score = destinations.size();
    return trail_score;
  };

  for (auto r : sv::iota(0UL, rows)) {
    for (auto c : sv::iota(0UL, cols)) {
      if (topo[r][c] == 0) {
        score += trace_path(r, c);
      }
    }
  }
  println("{}", score);
}

void part2(std::vector<std::vector<int>> const& topo) {
  const auto rows = topo.size();
  const auto cols = topo[0].size();
  std::size_t score = 0;

  auto trace_path = [&](std::size_t r, std::size_t c) {
    std::size_t trail_score = 0;
    std::vector<std::pair<int, std::pair<int, int>>> trails;

    trails.push_back({0, {r, c}});

    while (!trails.empty()) {
      auto [trail_val, trail_pos] = trails.back();
      trails.pop_back();
      auto [tr, tc] = trail_pos;
      for (auto [dr, dc] : directions) {
        auto nr = tr + dr;
        auto nc = tc + dc;
        if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) {
          continue;
        }
        if (trail_val == 8 && topo[nr][nc] == 9) {
          trail_score++;
          continue;
        }
        if (topo[nr][nc] == trail_val + 1) {
          trails.push_back({trail_val + 1, {nr, nc}});
        }
      }
    }
    return trail_score;
  };

  for (auto r : sv::iota(0UL, rows)) {
    for (auto c : sv::iota(0UL, cols)) {
      if (topo[r][c] == 0) {
        score += trace_path(r, c);
      }
    }
  }
  println("{}", score);
}

int main() {
  using namespace std::literals;
  auto lines = read_input("day10/in.txt");
  auto topo = flux::ref(lines)
                  .map([](std::string const& line) {
                    return flux::ref(line)
                        .map([](char c) { return c - '0'; })
                        .to<std::vector>();
                  })
                  .to<std::vector>();
  // println("{}", topo);
  MEASURE(part1(topo))
  MEASURE(part2(topo))
}

File: aoc-2024-rs/src/bin/day10.rs
use std::{collections::HashSet, vec};

const DIRECTIONS: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];

fn part1(topo: &Vec<Vec<u32>>) {
    let rows = topo.len();
    let cols = topo[0].len();

    let trace_path = |r: usize, c: usize| {
        let mut trails = vec![];
        let mut destinations = HashSet::new();
        let mut visited = vec![vec![false; cols]; rows];

        trails.push((0, (r as i32, c as i32)));

        while !trails.is_empty() {
            let (trail_val, (tr, tc)) = trails.pop().unwrap();
            visited[tr as usize][tc as usize] = true;
            for (dr, dc) in DIRECTIONS.iter() {
                let nr = tr + dr;
                let nc = tc + dc;
                if nr < 0 || nr >= rows as i32 || nc < 0 || nc >= cols as i32 {
                    continue;
                }
                if visited[nr as usize][nc as usize] {
                    continue;
                }
                if trail_val == 8 && topo[nr as usize][nc as usize] == 9 {
                    destinations.insert((nr, nc));
                    continue;
                }
                if topo[nr as usize][nc as usize] == trail_val + 1 {
                    trails.push((trail_val + 1, (nr, nc)));
                }
            }
        }
        destinations.len()
    };

    let mut score = 0;
    for r in 0..rows {
        for c in 0..cols {
            if topo[r][c] == 0 {
                score += trace_path(r, c);
            }
        }
    }
    println!("{:?}", score);
}

fn part2(topo: &Vec<Vec<u32>>) {
    let rows = topo.len();
    let cols = topo[0].len();

    let trace_path = |r: usize, c: usize| {
        let mut trails = vec![];
        let mut trail_score = 0;

        trails.push((0, (r as i32, c as i32)));

        while !trails.is_empty() {
            let (trail_val, (tr, tc)) = trails.pop().unwrap();
            for (dr, dc) in DIRECTIONS.iter() {
                let nr = tr + dr;
                let nc = tc + dc;
                if nr < 0 || nr >= rows as i32 || nc < 0 || nc >= cols as i32 {
                    continue;
                }
                if trail_val == 8 && topo[nr as usize][nc as usize] == 9 {
                    trail_score += 1;
                    continue;
                }
                if topo[nr as usize][nc as usize] == trail_val + 1 {
                    trails.push((trail_val + 1, (nr, nc)));
                }
            }
        }
        trail_score
    };

    let mut score = 0;
    for r in 0..rows {
        for c in 0..cols {
            if topo[r][c] == 0 {
                score += trace_path(r, c);
            }
        }
    }
    println!("{:?}", score);
}

fn main() {
    let input = include_str!("../../input/day10/in.txt");
    let topo: Vec<_> = input
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| c.to_digit(10).unwrap())
                .collect::<Vec<_>>()
        })
        .collect();
    // println!("{:?}", topo);
    part1(&topo);
    part2(&topo);
}
